jumanji/environments/routing/lbf/observer.py

```diff
@@ -0,0 +1,422 @@
+        agent_xs = jnp.where(visible_agents, transformed_positions[:, 0], -1)
+        agent_ys = jnp.where(visible_agents, transformed_positions[:, 1], -1)
+        agent_levels = jnp.where(visible_agents, all_agents.level, 0)
+
+        # Remove the current agent's info from all agent's infos.
+        agent_i_index = jnp.where(agent.id == all_agents.id, size=1)
+        agent_i_infos = jnp.array(
+            [
+                agent_xs[agent_i_index],
+                agent_ys[agent_i_index],
+                agent_levels[agent_i_index],
+            ]
+        ).ravel()
+
+        other_agents_indices = jnp.where(
+            agent.id != all_agents.id, size=self.num_agents - 1
+        )
+        agent_xs = agent_xs[other_agents_indices]
+        agent_ys = agent_ys[other_agents_indices]
+        agent_levels = agent_levels[other_agents_indices]
+
+        return agent_i_infos, agent_xs, agent_ys, agent_levels
+
+    def make_agents_view(self, agent: Agent, state: State) -> chex.Array:
+        """
+        Make an observation for a single agent based on the current state of the environment.
+        Returns the observation for the given agent."""
+
+        #  Check which agents within in the fov of the current agent.
+        visible_agents = jnp.all(
+            jnp.abs(agent.position - state.agents.position) <= self.fov, axis=-1
+        )
+
+        # Check which food items are visible and are not eaten.
+        visible_foods = (
+            jnp.all(
+                jnp.abs(agent.position - state.food_items.position) <= self.fov,
+                axis=-1,
+            )
+            & ~state.food_items.eaten
+        )
+
+        # Placeholder observation.
+        init_vals = jnp.array([-1, -1, 0])
+        agent_view = jnp.tile(init_vals, self.num_food + self.num_agents)
+
+        food_xs, food_ys, food_levels = self.extract_foods_info(
+            agent, visible_foods, state.food_items
+        )
+        agent_i_infos, agent_xs, agent_ys, agent_levels = self.extract_agents_info(
+            agent, visible_agents, state.agents
+        )
+
+        # Assign the foods and agents infos.
+        agent_view = agent_view.at[jnp.arange(0, 3 * self.num_food, 3)].set(food_xs)
+        agent_view = agent_view.at[jnp.arange(1, 3 * self.num_food, 3)].set(food_ys)
+        agent_view = agent_view.at[jnp.arange(2, 3 * self.num_food, 3)].set(food_levels)
+
+        # Always place the current agent's info first.
+        agent_view = agent_view.at[
+            jnp.arange(3 * self.num_food, 3 * self.num_food + 3)
+        ].set(agent_i_infos)
+
+        start_idx = 3 * self.num_food + 3
+        end_idx = start_idx + 3 * (self.num_agents - 1)
+        agent_view = agent_view.at[jnp.arange(start_idx, end_idx, 3)].set(agent_xs)
+        agent_view = agent_view.at[jnp.arange(start_idx + 1, end_idx, 3)].set(agent_ys)
+        agent_view = agent_view.at[jnp.arange(start_idx + 2, end_idx, 3)].set(
+            agent_levels
+        )
```

Comment:
For all of these the indices are unique and sorted, can you add those arguments here, as suggested in `utils.py` might give a small speed boost

Reply:
Yup :+1: